### PR TITLE
Component test improvements

### PR DESCRIPTION
**Pull request by bartoszmajsak**
This PR brings tests for "Connected Accounts Component" and introduces `TestContext` which hides all the boilerplate related to setting up Angular `TestBed`. We could start using it as a template/basis for all Angular component tests.

### Key features

- Each test ends with destroying the fixture (so we can catch errors like the one with the Profile Page in the future)
- Gives type-safe way of interacting with both component under test and host component (one used from within the test as an example usage)
- `TestContext` is a single point of maintenance so any changes in Angular component testing can be addressed in the single class. Can also serve as evolving utility class to simplify common interactions across the tests

### CODE REVIEW

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-1 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Comment by bartoszmajsak** [comment]
While I was working on this test I noticed there is quite a bit of duplication in the test code, but I also wanted to make sure that we catch errors like that, occurring in `ngDestroy`, for every test.

I found a [pretty good article](https://medium.com/claritydesignsystem/angular-testing-made-easy-4e11f6044129) showing the concept of `TestContext` and I believe this is a good foundation to move forward. If you have time you can also [watch a talk](https://www.youtube.com/watch?v=9hrPqWETPao) by the same author.

Adding that many reviewers to share it and discuss.

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-2 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Comment by bartoszmajsak** [comment]
Another idea worth consider is to bring type-safe mocking instead of what we have available out-of-the-box from Jasmine, e.g https://github.com/NagRock/ts-mockito

**Comment by bartoszmajsak** [comment]
Any thoughts around that? I see more PRs coming with the tests so I was wondering if it would be worth bringing this in or ditch if does not make sense.

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-3 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

File: src/testing/test-context.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     * Jasmine creates plain objects and modifying their prototype is definitely a bad idea.
+     * So until we find better way..
+     */
+    Object.assign(this, TestContext.prototype);
+  });
+
+  beforeEach(async(function(this: TestContext<T, H>) {
+    const declarations = [ testedType, hostType ];
+    if (moduleMetadata && moduleMetadata.declarations) {
+      declarations.push(...moduleMetadata.declarations);
```
Comment (sbose78) [inline src/testing/test-context.ts]: Cool.

File: src/testing/test-context.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    TestBed.configureTestingModule({...moduleMetadata, declarations: declarations })
+      .compileComponents();
+  }));
+
+  beforeEach(function(this: TestContext<T, H>) {
+    this.fixture = TestBed.createComponent(hostType);
+    this.fixture.detectChanges();
```
Comment (sbose78) [inline src/testing/test-context.ts]: Why do we need this here ?
Doesn't harm though..

File: src/testing/test-context.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  beforeEach(async(function(this: TestContext<T, H>) {
+    const declarations = [ testedType, hostType ];
+    if (moduleMetadata && moduleMetadata.declarations) {
+      declarations.push(...moduleMetadata.declarations);
+    }
+    TestBed.configureTestingModule({...moduleMetadata, declarations: declarations })
+      .compileComponents();
+  }));
+
+  beforeEach(function(this: TestContext<T, H>) {
```
Comment (sbose78) [inline src/testing/test-context.ts]: In the `src/app/profile/overview/connected-accounts/connected-accounts.component.spec.ts` you added, is this the method that is actually getting called ?

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-4 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

File: src/testing/test-context.ts (hunk 2 above)
Comment (bartoszmajsak) [reply src/testing/test-context.ts]: Yeah, it's not a big overhead, but also might be needed to reflect all the changes in the DOM I believe.

File: src/testing/test-context.ts (hunk 3 above)
Comment (bartoszmajsak) [reply src/testing/test-context.ts]: Every `beforeEach` is registered in Jasmine runner, so yeah, it is called by the test framework.

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-5 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-6 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Comment by corinnekrych** [comment]
I think the `initContext` and `TestContext` to embed inside jasmine's context the angular TestBed boiler plate sounds like a good idea to me. I'm ready to try it on the component/service tests I wrote!
 One note though: rather than including the source code `testing\test-context.ts` I think would be great in a separate lib published to npm.

**Comment by bartoszmajsak** [comment]
> One note though: rather than including the source code testing\test-context.ts I think would be great in a separate lib published to npm.

Let's give it a try in this project and evolve based on our needs and once settled (hopefully after your try-out) I'm strongly for that, especially if we can keep it generic.

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-7 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Comment by corinnekrych** [comment]
@bartoszmajsak pls rebase and we can merge then!

**Comment by fabric8cd** [comment]
@bartoszmajsak fabric8/fabric8-ui:SNAPSHOT-PR-2234-8 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2234-fabric8-ui.badger.fabric8.io

**Pull request merged**